lib/smb2/packet.rb

```diff
@@ -132,44 +145,56 @@ def self.data_buffer(name, bit_length=16, opts={})
     self.unsigned "#{name}_padding", opts[:padding] if opts.has_key?(:padding)
     self.unsigned "#{name}_length", bit_length, endian: 'little'
 
-    class_eval do
-
-      define_method(name) do
-        field_offset = self.send("#{name}_offset")
-        field_length = self.send("#{name}_length")
-        # Must use #to_s so we get the whole packet packed because offset is from
-        # beginning of header.
-        to_s.slice(field_offset, field_length)
-      end
-
-      define_method("#{name}=") do |other|
-        @data_buffers[name] = other
-        recalculate
-      end
+    define_method(name) do
+      field_offset = self.send("#{name}_offset")
+      field_length = self.send("#{name}_length")
+      # Must use #to_s so we get the whole packet packed because offset is from
+      # beginning of header.
+      to_s.slice(field_offset, field_length)
+    end
 
+    define_method("#{name}=") do |other|
+      @data_buffers[name] = other
+      recalculate
     end
 
     self
   end
 
+  ##
+  # Instance methods
+  ##
+
   # @see BitStruct#initialize
   def initialize(*args)
     @data_buffers = {}
     super do
-      if !self.class.data_buffer_fields.empty?
-        self.class.data_buffer_fields.each do |buffer_name|
+      if !data_buffer_fields.empty?
+        data_buffer_fields.each do |buffer_name|
           @data_buffers[buffer_name] = self.send(buffer_name) || ""
         end
         recalculate
       end
       yield self if block_given?
     end
+
+    if respond_to?(:header) && self.class.respond_to?(:command)
```

Comment:
I actually started implementing it that way originally and went to a method instead.  I don't have a good reason why.